win32ss/user/ntuser/class.c

```diff
@@ -1275,39 +1290,39 @@ IntGetAtomFromStringOrAtom(
     if (ClassName->Length != 0)
     {
         WCHAR szBuf[65];
-        PWSTR AtomName;
-        NTSTATUS Status;
+        PWSTR AtomName = szBuf;
+        NTSTATUS Status = STATUS_INVALID_PARAMETER;
 
         *Atom = 0;
 
         /* NOTE: Caller has to protect the call with SEH! */
-
-        if (ClassName->Length != 0)
+        if (ClassName->Length / sizeof(WCHAR) >= ARRAYSIZE(szBuf))
         {
-            /* FIXME: Don't limit to 64 characters! use SEH when allocating memory! */
-            if (ClassName->Length / sizeof(WCHAR) >= sizeof(szBuf) / sizeof(szBuf[0]))
+            AtomName = ExAllocatePoolWithTag(PagedPool,
+                                             ClassName->Length + sizeof(UNICODE_NULL),
+                                             TAG_STRING);
+            if (AtomName == NULL)
             {
-                EngSetLastError(ERROR_INVALID_PARAMETER);
-                return (RTL_ATOM)0;
+                EngSetLastError(ERROR_OUTOFMEMORY);
+                return FALSE;
             }
-
-            /* We need to make a local copy of the class name! The caller could
-               modify the buffer and we could overflow in RtlLookupAtomInAtomTable.
-               We're protected by SEH, but the ranges that might be accessed were
-               not probed... */
-            RtlCopyMemory(szBuf,
-                          ClassName->Buffer,
-                          ClassName->Length);
-            szBuf[ClassName->Length / sizeof(WCHAR)] = UNICODE_NULL;
-            AtomName = szBuf;
         }
-        else
-            AtomName = ClassName->Buffer;
+
+        /* We need to make a local copy of the class name! The caller could
+           modify the buffer and we could overflow in RtlLookupAtomInAtomTable.
+           We're protected by SEH, but the ranges that might be accessed were
```

Comment:
Where is SEH? The caller? Or is the comment outdated?

Reply:
Yeah I'm still confused about what data is trusted and what data is untrusted.
If this function processes untrusted data, then it _should_ have a `finally` handler to free the buffer.

The usual process is to just make the "safe" copy in the top level `NtUser` function... I'm not sure why things seem to be so complicated here.

Reply:
I had a finally handler, but timo said gcc's seh is too unreliable for that.

Reply:
Sad.

Reply:
To be clear: this function should not process untrusted data.